pkg/input/arm/parser.go

```diff
@@ -0,0 +1,186 @@
+// limitations under the License.
+
+package arm
+
+import (
+	"errors"
+	"fmt"
+)
+
+func parse(tokens []token) (expression, error) {
+	p := &parser{remaining: tokens}
+	return p.parse()
+}
+
+type parser struct {
+	remaining []token
+}
+
+func (p *parser) parse() (expression, error) {
+	if p.done() {
+		return nil, newParserError(errors.New("can't build expression from 0 tokens"))
+	}
+
+	tkn := p.pop()
+
+	// We may need to add support for more types (integer, bool, arrays and
+	// objects) when we add support for more functions.
+	// https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/template-expressions
+	if strToken, ok := tkn.(stringLiteral); ok {
+		return stringLiteralExpr(strToken), nil
+	}
+
+	// If we reach here, we are building a function expression, because there are
+	// no "direct" identifier dereferences in ARM template expressions. The
+	// identifier is the function name.
+	idToken, ok := tkn.(identifier)
+	if !ok {
+		return nil, newParserError(fmt.Errorf("expected token %#v to be an identifier", tkn))
+	}
+
+	// expect an open paren
+	if p.done() {
+		return nil, newParserError(errors.New("expression cannot terminate with an identifier"))
+	}
+	tkn = p.pop()
+	if _, ok := tkn.(openParen); !ok {
+		return nil, newParserError(fmt.Errorf("expected token %#v to be a paren", tkn))
+	}
+	if p.done() {
+		return nil, newParserError(errors.New("expression cannot terminate with an open paren"))
+	}
+
+	var args []expression
+	for {
+		tkn := p.peek()
+		if _, ok := tkn.(closeParen); ok {
+			p.pop() // pop the close paren
+			expr := functionExpr{name: string(idToken), args: args}
+			if p.done() {
+				return expr, nil
+			}
+			if _, ok := p.peek().(dot); ok {
+				return p.buildPropertyAccessExpression(expr)
+			}
+			return expr, nil
+		}
+
+		// There is a comma between args, so not before the first arg
+		if len(args) > 0 {
+			if _, ok := p.peek().(comma); !ok {
+				return nil, newParserError(fmt.Errorf("expected token %#v to be a comma", tkn))
+			}
+			p.pop() // pop the comma
+		}
+
+		nextArg, err := p.parse()
+		if err != nil {
+			return nil, err
+		}
+		args = append(args, nextArg)
+	}
+}
+
+func (p *parser) buildPropertyAccessExpression(expr expression) (expression, error) {
+	// we only enter this function from parse() if we peeked at a dot, so we know
+	// it gets past here at least once, and so always builds a real property
+	// access expression.
+	if p.done() {
+		return expr, nil
+	}
+	if _, ok := p.peek().(dot); !ok {
+		return expr, nil
+	}
+
+	p.pop() // pop the dot
+	if p.done() {
+		return nil, newParserError(errors.New("expression cannot terminate with a dot"))
+	}
+	tkn := p.pop()
+	nextPropChainElement, ok := tkn.(identifier)
+	if !ok {
+		return nil, newParserError(fmt.Errorf("expected token %#v to be an identifier", tkn))
+	}
+	expr = propertyExpr{obj: expr, property: string(nextPropChainElement)}
+	return p.buildPropertyAccessExpression(expr)
+}
+
+func (p *parser) peek() token {
```

Comment:
Nitpick: I think it would be slightly cleaner to return `(token, bool)` here and in `pop` like the `tokenizer` does, and maybe remove `done()`?.  That forces the programmer to "think" about the boolean that's returned, whereas it's a bit easier to forget calling `done()`.

Alternatively, we could also return `nil` if we're done but that seems a bit nastier...

Reply:
Agreed, let's use the same API for both. I'll change this.